src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -979,6 +969,33 @@ namespace cryptonote
     CHECK_AND_ASSERT_MES(rc == 0, false, "failed to convert ed25519 pubkey to x25519");
     crypto_sign_ed25519_sk_to_curve25519(keys.key_x25519.data, keys.key_ed25519.data);
 
+    // Legacy primary SN key file; we only load this if it exists, otherwise we use `key_ed25519`
+    // for the primary SN keypair.  (This key predates the Ed25519 keys and so is needed for
+    // backwards compatibility with existing active service nodes.)  The legacy key consists of
+    // *just* the private point, but not the seed, and so cannot be used for full Ed25519 signatures
+    // (which rely on the seed for signing).
+    if (m_service_node) {
+      if (!epee::file_io_utils::is_file_exist(m_config_folder + "/key")) {
+        epee::wipeable_string privkey_signhash;
+        privkey_signhash.resize(64);
+        unsigned char* pk_sh_data = reinterpret_cast<unsigned char*>(privkey_signhash.data());
+        crypto_hash_sha512(pk_sh_data, keys.key_ed25519.data, 32);
+        pk_sh_data[0] &= 248;
+        pk_sh_data[31] &= 63;
+        pk_sh_data[31] |= 64;
```

Comment:
We need some reference of where this was pulled from. In Sodium this is where it's done

https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_sign/ed25519/ref10/keypair.c#L70

But they do 

```cpp
    crypto_hash_sha512(h, ed25519_sk, 32);
    h[0] &= 248;
    h[31] &= 127;
    h[31] |= 64;
```

I've found it in our code-base in the reference crypto implementations, would be good to just drop a comment to that. 

Though, its slightly confusing in that we amalgamate the reference implementation into `crypto-ops` which doesn't include this signing code. Has it been verified that these changes are sufficient for a legacy key owner to validate a signature from a post v8.x SN and vice versa?

Reply:
Those two bitwise operations are identical; the only difference is whether or not it clears the 7th bit or not before it explicitly sets it.

Since all this is doing is changing the private key value (to prevent small subgroup attacks), and Monero keys are unrestrained values, this will always be a valid Monero private key point, but many Monero key points won't be possible with this generation.

Reply:
Ah yeah nice, they do work out to be the same.